LucenePlusPlus failed to build on Boost 1.58.0

Here's a build  log on system with latest boost:
[ 15%] Building CXX object src/core/CMakeFiles/lucene++.dir/search/NumericRangeQuery.cpp.o
In file included from /usr/include/boost/iterator/iterator_categories.hpp:22:0,
                 from /usr/include/boost/iterator/iterator_facade.hpp:14,
                 from /usr/include/boost/filesystem/path.hpp:28,
                 from /usr/include/boost/filesystem/fstream.hpp:21,
                 from /tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/Lucene.h:22,
                 from /tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/include/LuceneInc.h:20,
                 from /tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/build/src/core/cotire/lucene++_CXX_prefix.cxx:4,
                 from /tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/build/src/core/cotire/lucene++_CXX_prefix.hxx:4:
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = std::basic_string<wchar_t>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = std::basic_string<wchar_t>&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = std::basic_string<wchar_t>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = std::basic_string<wchar_t>&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = std::basic_string<wchar_t>; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:49:57:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
     BOOST_STATIC_ASSERT_MSG(
     ^
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<unsigned char>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<unsigned char>&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<unsigned char>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<unsigned char>&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = Lucene::Collection<unsigned char>; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:62:50:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<int>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<int>&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<int>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<int>&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = Lucene::Collection<int>; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:65:50:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<long long int>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<long long int>&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<long long int>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<long long int>&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = Lucene::Collection<long long int>; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:68:50:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<double>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<double>&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<double>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<double>&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = Lucene::Collection<double>; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:71:49:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<std::basic_string<wchar_t> >; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<std::basic_string<wchar_t> >&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = Lucene::Collection<std::basic_string<wchar_t> >; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = Lucene::Collection<std::basic_string<wchar_t> >&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = Lucene::Collection<std::basic_string<wchar_t> >; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:74:49:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
/usr/include/boost/variant/get.hpp: In instantiation of ‘typename boost::add_reference<T>::type boost::strict_get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = boost::shared_ptr<Lucene::LuceneObject>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = boost::shared_ptr<Lucene::LuceneObject>&]’:
/usr/include/boost/variant/get.hpp:284:33:   required from ‘typename boost::add_reference<T>::type boost::get(boost::variant<T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19>&) [with U = boost::shared_ptr<Lucene::LuceneObject>; T0 = int; T1 = long long int; T2 = double; T3 = boost::blank; T4 = boost::detail::variant::void_; T5 = boost::detail::variant::void_; T6 = boost::detail::variant::void_; T7 = boost::detail::variant::void_; T8 = boost::detail::variant::void_; T9 = boost::detail::variant::void_; T10 = boost::detail::variant::void_; T11 = boost::detail::variant::void_; T12 = boost::detail::variant::void_; T13 = boost::detail::variant::void_; T14 = boost::detail::variant::void_; T15 = boost::detail::variant::void_; T16 = boost::detail::variant::void_; T17 = boost::detail::variant::void_; T18 = boost::detail::variant::void_; T19 = boost::detail::variant::void_; typename boost::add_reference<T>::type = boost::shared_ptr<Lucene::LuceneObject>&]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:25:65:   required from ‘static TYPE Lucene::VariantUtils::get(VAR) [with TYPE = boost::shared_ptr<Lucene::LuceneObject>; VAR = boost::variant<int, long long int, double, boost::blank>]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/include/VariantUtils.h:77:44:   required from ‘static int32_t Lucene::VariantUtils::hashCode(VAR) [with VAR = boost::variant<int, long long int, double, boost::blank>; int32_t = int]’
/tmp/sbopkg.zKrhQy/LucenePlusPlus-rel_3.0.7/src/core/search/NumericRangeQuery.cpp:175:43:   required from here
/usr/include/boost/variant/get.hpp:178:5: error: invalid application of ‘sizeof’ to incomplete type ‘boost::STATIC_ASSERTION_FAILURE<false>’
make[2]: *** [src/core/CMakeFiles/lucene++.dir/search/NumericRangeQuery.cpp.o] Error 1
make[1]: *** [src/core/CMakeFiles/lucene++.dir/all] Error 2
make: *** [all] Error 2

--- a/include/VariantUtils.h
+++ b/include/VariantUtils.h
@@ -8,6 +8,7 @@
 #define VARIANTUTILS_H
 
 #include <boost/any.hpp>
+#include <boost/version.hpp>
 #include "Lucene.h"
 #include "MiscUtils.h"
 
@@ -22,7 +23,11 @@ public:
 
     template <typename TYPE, typename VAR>
     static TYPE get(VAR var) {
+#if BOOST_VERSION < 105800
         return var.type() == typeid(TYPE) ? boost::get<TYPE>(var) : TYPE();
+#else
+        return var.type() == typeid(TYPE) ? boost::relaxed_get<TYPE>(var) : TYPE();
+#endif
     }
 
     template <typename TYPE, typename VAR>

